src/sync/client-gc.ts

```diff
@@ -0,0 +1,31 @@
+import type {ClientID} from './client-id';
+import type * as dag from '../dag/mod';
+import {getClients, setClients} from './clients';
+
+const CLIENT_MAX_INACTIVE_IN_MS = 7 * 24 * 60 * 60 * 1000; // 7 days
+const GC_INTERVAL_MS = 5 * 60 * 1000; // 5 minutes
+
+export function initClientGC(
+  clientID: ClientID,
+  dagStore: dag.Store,
+): () => void {
+  const intervalID = window.setInterval(async () => {
+    await dagStore.withWrite(async (write: dag.Write) => {
+      const clients = await getClients(write);
+      const now = Date.now();
+      const clientsAfterGC = Array.from(clients).filter(
+        ([id, client]) =>
+          id === clientID /* never collect ourself */ ||
+          now - client.heartbeatTimestampMs <= CLIENT_MAX_INACTIVE_IN_MS,
+      );
+      if (clientsAfterGC.length !== clients.size) {
```

Comment:
You could have kept the old code and just add a flag.

```js
let changed = false;
...
changed = true;
...
if (changed) {
  ...
}
```

Anyway, the current code works and is not that much complicated

Reply:
yes I considered that as well but decided I liked this functional programming version a tiny bit better.